payjoin-directory/src/lib.rs

```diff
@@ -112,21 +118,52 @@ impl<D: Db> Service<D> {
         Ok(())
     }
 
-    pub async fn serve_tcp(self, listener: tokio::net::TcpListener) -> Result<(), BoxError> {
-        while let Ok((stream, _)) = listener.accept().await {
-            let io = TokioIo::new(stream);
-            let service = self.clone();
-            tokio::spawn(async move {
-                service.metrics.record_connection();
-                if let Err(err) =
-                    http1::Builder::new().serve_connection(io, service).with_upgrades().await
-                {
-                    error!("Error serving connection: {:?}", err);
+    #[cfg(feature = "acme")]
+    pub async fn serve_acme<EC, EA>(self, listener: TcpListener, acme_config: AcmeConfig<EC, EA>)
+    where
+        EC: 'static + std::fmt::Debug,
+        EA: 'static + std::fmt::Debug,
+    {
+        let tcp_incoming = TcpListenerStream::new(listener);
+
+        let tls_incoming = acme_config.incoming(tcp_incoming, Vec::new());
+
+        self.serve_connections(tls_incoming).await;
+    }
+
+    pub async fn serve_tcp(self, listener: TcpListener) {
+        let tcp_incoming = TcpListenerStream::new(listener);
+        self.serve_connections(tcp_incoming).await;
+    }
+
+    async fn serve_connections<S, I>(self, mut incoming_connections: S)
+    where
+        S: Stream<Item = tokio::io::Result<I>> + Unpin,
+        I: tokio::io::AsyncRead + tokio::io::AsyncWrite + Send + Unpin + 'static,
+    {
+        while let Some(conn) = incoming_connections.next().await {
+            match conn {
+                Ok(stream) => {
+                    let service = self.clone();
+                    tokio::spawn(async move { service.serve_connection(stream).await });
                 }
-            });
+                Err(err) => {
+                    error!("Accept error: {err}")
+                }
+            }
         }
+    }
 
-        Ok(())
+    // TODO https://docs.rs/tower/0.4.13/tower/make/trait.MakeService.html
+    async fn serve_connection<I>(&self, stream: I)
+    where
+        I: tokio::io::AsyncRead + tokio::io::AsyncWrite + Send + Unpin + 'static, // FIXME why static?
```

Comment:
istm connections are handled on their own spawned taks which can run indefinitely and must be `'static` because they can outlive the serve_connection function, the runtime doesn't know when they'll complete, & they can run concurrently with other tasks. Seems like a constraint from .with_upgrades().